grpc/federation/cel/list.go

```diff
@@ -61,6 +110,93 @@ func makeFirst(mef cel.MacroExprFactory, target ast.Expr, args []ast.Expr) (ast.
 	return mef.NewCall(operators.OptIndex, filter, mef.NewLiteral(types.Int(0))), nil
 }
 
+func makeSortAsc(mef cel.MacroExprFactory, target ast.Expr, args []ast.Expr) (ast.Expr, *cel.Error) {
+	return makeSort(listSortAscFunc, mef, target, args)
+}
+
+func makeSortDesc(mef cel.MacroExprFactory, target ast.Expr, args []ast.Expr) (ast.Expr, *cel.Error) {
+	return makeSort(listSortDescFunc, mef, target, args)
+}
+
+func makeSortStableAsc(mef cel.MacroExprFactory, target ast.Expr, args []ast.Expr) (ast.Expr, *cel.Error) {
+	return makeSort(listSortStableAscFunc, mef, target, args)
+}
+
+func makeSortStableDesc(mef cel.MacroExprFactory, target ast.Expr, args []ast.Expr) (ast.Expr, *cel.Error) {
+	return makeSort(listSortStableDescFunc, mef, target, args)
+}
+
+func makeSort(function string, mef cel.MacroExprFactory, target ast.Expr, args []ast.Expr) (ast.Expr, *cel.Error) {
+	mp, err := parser.MakeMap(mef, target, args)
+	if err != nil {
+		return nil, err
+	}
+	return mef.NewCall(function, target, mp), nil
+}
+
+func sortAsc(orig, expanded ref.Val) ref.Val {
+	return sortRefVal(orig, expanded, -types.IntOne, false)
+}
+
+func sortDesc(orig, expanded ref.Val) ref.Val {
+	return sortRefVal(orig, expanded, types.IntOne, false)
+}
+
+func sortStableAsc(orig, expanded ref.Val) ref.Val {
+	return sortRefVal(orig, expanded, -types.IntOne, true)
+}
+
+func sortStableDesc(orig, expanded ref.Val) ref.Val {
+	return sortRefVal(orig, expanded, types.IntOne, true)
+}
+
+func sortRefVal(orig, expanded ref.Val, direction types.Int, stable bool) ref.Val {
+	origLister := orig.(traits.Lister)
+	expandedLister := expanded.(traits.Lister)
+
+	// The element being sorted must be values before expansion.
+	type sortVal struct {
+		OrigVal     ref.Val
+		ExpandedVal ref.Val
+	}
+	vals := make([]*sortVal, 0, int64(origLister.Size().(types.Int)))
+	for i := types.IntZero; i < origLister.Size().(types.Int); i++ {
+		origVal := origLister.Get(i)
+		expVal := expandedLister.Get(i)
+		if _, ok := expVal.(traits.Comparer); !ok {
```

Comment:
👍

Reply:
I initially thought this method would be fine, but I realized that with this implementation, it cannot result in an error during the compile phase. Please add tests in `validator/validator_test.go` to check if it can produce an error